### ISSUE TITLE
Tensor min max over axis crashes

### ISSUE THREAD

**Issue by cemde** [pytorch/pytorch#90412]
### 🐛 Describe the bug

`torch.Tensor.max()` and `torch.Tensor.min()` work fine when when getting the overall extremum, but as soon as I specify the axis, e.g. `torch.Tensor.max(dim=-1)`, the script crashes quietly. Importantly, there is NO error thrown. Weirdly the script works about 20% of the time and doesnt 80% of the time. Rebooting the OS doesn't work.

```python
import itertools
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.autograd import Variable
import functorch as FT
import functools

# create data
torch.manual_seed(0)
TAU = 0.01
P = torch.softmax(torch.randn(10,3), dim=-1)
Y = torch.distributions.Categorical(P).sample().unsqueeze(dim=-1)

def func(f: torch.Tensor, y: torch.Tensor, tau: float) -> torch.Tensor:
    f_true = f[torch.arange(f.shape[0]), y]
    print(f.max())
    f_pred = f.max(dim=-1)
    print("NOT CRASHED :D")


c_out = func(P, Y, TAU)
print(c_out)
```

This runs well and prints `tensor(0.7848)` then crashes. The same behaviour can be observed for `torch.max` and `torch.min`.

### Versions

PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:38:35) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.1
[pip3] pytorch-lightning==1.7.2
[pip3] torch==1.13.0
[pip3] torchaudio==0.13.0
[pip3] torchinfo==1.7.0
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.14.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.4.0           hecd8cb5_637  
[conda] mkl-service               2.4.0           py310hca72f7f_0  
[conda] mkl_fft                   1.3.1           py310hf879493_0  
[conda] mkl_random                1.2.2           py310hc081a56_0  
[conda] numpy                     1.23.1          py310hdcd3fac_0  
[conda] numpy-base                1.23.1          py310hfd2de13_0  
[conda] pytorch-lightning         1.7.2                    pypi_0    pypi
[conda] torch                     1.13.0                   pypi_0    pypi
[conda] torchaudio                0.13.0                   pypi_0    pypi
[conda] torchinfo                 1.7.0                    pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchvision               0.14.0                   pypi_0    pypi

**Comment by cemde** [comment]
I have had a lot of issues with torch and numpy on MacOS in the past. It just takes me a few hours everytime distilling my very long code into a minimum broken script and then checking if numpy is the culpit, because it was imported somewhere along the way.

**Comment by albanD** [comment]
Does importing torch before numpy solves the problem?

**Comment by cemde** [comment]
> 

Importing numpy after torch also works

**Comment by albanD** [comment]
That most likely points to ABI issue then when numpy exposes some symbols globally and so PyTorch doesn't reload the version of the lib it needs.

cc @rgommers is there any issue/NEP we can can use to discuss this question?

**Comment by rgommers** [comment]
I'd say this issue is as good as any for now; a corresponding NumPy issue can be opened if we determine it's a bug in NumPy. What I see in the reproducer though is that `numpy` is installed with Conda, and `torch` with Pip. This is a very bad idea, and tends to lead to issues with incompatible BLAS/LAPACK/OpenMP (as @vadimkantorov also pointed out above). 

@cemde if you install all packages in your environment with Conda, does the problem go away? If so, what if you install everything with Pip?

> `torch.Tensor.max()` and `torch.Tensor.min()` work fine when when getting the overall extremum, but as soon as I specify the axis, e.g. `torch.Tensor.max(dim=-1)`, the script crashes quietly.

My guess here is that `max()` is a global reduction and hence single-threaded, while `max(dim=-1)` is a set of independent reductions along one axis so gets parallelized with OpenMP. The OpenMP version used in the `torch`wheel from PyPI may conflict with the Intel OpenMP that gets pulled in via MKL with Conda.

**Comment by vadimkantorov** [comment]
@rgommers Is it possible to somehow test for this? E.g. have a test that loads only numpy, enumerates loaded shared libraries and then prints out all exported symbols in them; and then loads only pytorch and does the same. Then we could look at the table intersecting two sets, or also reason about compatible / incompatible BLAS/OpenMP versions.

One could even have this utility as a separate script (similar to collect_env.py) that could be used without a workable PyTorch installation

**Comment by rgommers** [comment]
> Is it possible to somehow test for this? E.g. have a test that loads only numpy, enumerates loaded shared libraries and then prints out all exported symbols in them; and then loads only pytorch and does the same.

I'm not aware of such a test having been written, but it should be possible. That said, it's nontrivial to implement symbol checks, e.g. if you have the same symbol in two shared libraries, it matters how it got loaded (`RTLD_LOCAL` vs `RTLD_GLOBAL` etc.). 

One easier improvement is for `collect_env.py` to check if `numpy` and `torch` were installed with the same package manager, and if not then print a large warning. I'd even be in favor of adding a warning in `torch/__init__.py` for the case of "one with pip, one with conda" in particular. We've had enough trouble with that to justify doing that.

> A related issue: gh-82098

Let me comment on that issue regarding symbol / shared library detection.

**Comment by vadimkantorov** [comment]
Maybe a recipe can be existed that does run `python3 -c "import torch"` and `python3 -c "import "import numpy"` under strace/LD_PRELOAD and trace how dlopen got called? (if that's how these libraries get loaded) or maybe linux loaders have some tracing env. variable?

**Comment by vadimkantorov** [comment]
As seems from LD_DEBUG=libs and ltrace (can work without ptrace-permissions?) might be helpful for this: https://unix.stackexchange.com/questions/448964/how-do-i-trace-library-loading (https://stackoverflow.com/a/65452060/445810)
